src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioActiveStatementProvider.cs

```diff
@@ -0,0 +1,118 @@
+                                var instructionMap = PooledDictionary<ActiveInstructionId, (DkmInstructionSymbol Symbol, ArrayBuilder<Guid> Threads, int Index, ActiveStatementFlags Flags)>.GetInstance();
+                                GroupActiveStatementsByInstructionId(instructionMap, activeStatementsResult.ActiveStatements);
+
+                                int pendingStatements = instructionMap.Count;
+                                builders[runtimeIndex] = ArrayBuilder<ActiveStatementDebugInfo>.GetInstance(pendingStatements);
+                                builders[runtimeIndex].Count = pendingStatements;
+
+                                foreach (var (instructionId, (symbol, threads, index, flags)) in instructionMap)
+                                {
+                                    symbol.GetSourcePosition(workList, DkmSourcePositionFlags.None, InspectionSession: null, sourcePositionResult =>
+                                    {
+                                        if (cancellationToken.IsCancellationRequested)
+                                        {
+                                            workList.Cancel();
+                                            completion.SetCanceled();
+                                        }
+
+                                        if (sourcePositionResult.ErrorCode == 0)
+                                        {
+                                            builders[runtimeIndex][index] = new ActiveStatementDebugInfo(
+                                                instructionId,
+                                                sourcePositionResult.SourcePosition.DocumentName,
+                                                ToLinePositionSpan(sourcePositionResult.SourcePosition.TextSpan),
+                                                threads.ToImmutableAndFree(),
+                                                flags);
+                                        }
+
+                                        // the last active statement of the current runtime has been processed:
+                                        if (Interlocked.Decrement(ref pendingStatements) == 0)
+                                        {
+                                            instructionMap.Free();
+
+                                            // the last active statement of the last runtime has been processed:
+                                            if (Interlocked.Decrement(ref pendingRuntimes) == 0)
+                                            {
+                                                completion.SetResult(builders.ToImmutableArrayAndFree());
+                                            }
+                                        }
+                                    });
+                                }
+                            });
+
+                            runtimeCount++;
```

Comment:
Might be easier to follow if this is up by line 43. I nearly thought you missed it and got lost in the flock of seagulls.

Reply:
Good point.